### ISSUE TITLE
S3 basic_external doesn't work

### ISSUE THREAD

**Issue by pauldraper** [datopian/giftless#106]
The signed S3 URL doesn't work for uploading.

Git LFS receives a 401 when uploading.

**Comment by devinnasar** [comment]
I believe this to be related to the configuration for signature version in the boto3 client here: https://github.com/datopian/giftless/blob/5b3f7204321e8775b9eac3953a4894988ad78b7c/giftless/storage/amazon_s3.py#L29

See this issue in botocore: https://github.com/boto/botocore/issues/2109. If the fix indicated at the bottom were implemented it may correct the issue.

I'm receiving a similar message: "The request signature we calculated does not match the signature you provided. Check your key and signing method." when I try to run `git push` after following the tutorial for google cloud storage, but modified to run against s3.

**Comment by athornton** [comment]
Feel free to try it and make a PR.  I'm at a conference this week but I can probably get to it next week (disclaimer: I don't work for Datopian, so my time to work on this is entirely at my management's discretion, so if stuff is blowing up, I won't, but at least right now it seems like I should be able to, since that looks like a very small change).

**Comment by devinnasar** [comment]
I'll give it a shot. I need to adjust my setup to try and use local giftless.

**Comment by rufuspollock** [comment]
@devinnasar did you get a PR in on this?

**Comment by rufuspollock** [comment]
@devinnasar @athornton any thoughts on this?

**Comment by athornton** [comment]
Yeah, I think I completely forgot this existed and utterly dropped the ball.  Sorry.

I should take a look again.

**Comment by Y-o-p** [comment]
**Interim Solution**

1. Start storing your aws credentials in `~/.aws/credentials` and `~/.aws/config` if you're not already. You can do this with `aws configure`
2. Add this S3 section and set it to s3v4. Here's my config for example:
```yaml
[default]
region = us-east-1 # Leave your original region
s3 =
    signature_version = s3v4
    addressing_style = virtual
```
amazon_s3.py doesn't explicitly construct a Config, so it'll fall back to using these options.

3. If you're using containers, mount your `~/.aws` with this option when starting the container:  `-v $HOME/.aws:/app/.aws:ro`

**Long term solution**

Like @devinnasar stated, update `self.s3_client = boto3.client("s3", endpoint_url=endpoint)` to explicitly set a Config signature version (s3v4). Or I think if you update boto, it should use the new signature version.
